### PR TITLE
Feature: Type level literals

### PR DESCRIPTION
**Pull request by zardyh**
**Note**: Only `string` and `int` kinds for now, enforced badly by the parser. Technically nothing's stopping us from having the other literals in the type level, except for the fact that they don't have associated classes.

This expands upon the previous work on promoting things by promoting the `int` and `string` types to also be kinds. The inhabitants of those kinds being, of course, uninhabited. It also adds our first named built-in module:

```
module Amc = begin
  class known_int (x : int) begin
    val int_value : int
  end
  class known_string (x : int) begin
    val int_value : string
  end
end
```

As a demonstration for how expressive this feature is, I've included a very small Servant-inspired example of a type-safe HTTP router thing.

### CODE REVIEW

**Comment by zardyh** [comment]
@SquidDev could you take care of the hlints?

File: src/Backend/Lua/Builtin.hs [hunk 1]
```diff
@@ -118,6 +118,30 @@ builtins =
           return var[1]
         end
       |] )
+  , ( vStrVal, "__builtin_strval", [], Nothing
```
Comment (SquidDev) [inline src/Backend/Lua/Builtin.hs]: Ew, ew, ew.

File: src/Syntax/Implicits.hs [hunk 2]
```diff
@@ -88,8 +90,28 @@ deriving instance (Show info, Show (Var p), Show (Ann p)) => Show (ImplicitScope
...
+  apply m (ManyMore is t) = ManyMore (apply m is) (apply m t)
+
+instance Ord (Var p) => Substitutable p (Implicit i p) where
+  ftv i = ftv (i ^. implType) Set.\\ boundByImpl i
+  apply m i = i & implHead %~ apply m' where
+    m' = m `Map.withoutKeys` boundByImpl i
+
 -- | Insert a choice for a *fully-known* (@Solved@) implicit parameter
--- (the variable @v@) of type @tau@ at the given trie.
+-- (the vriable @v@) of type @tau@ at the given trie.
```
Comment (SquidDev) [inline src/Syntax/Implicits.hs]: Can we revert this?

File: tests/parser/fail_span.ml [hunk 3]
```diff
@@ -1 +0,0 @@
-let main : "foo" = 1
```
Comment (SquidDev) [inline tests/parser/fail_span.ml]: Can we add this test back in, just with something else? `let main : match` will be good.

File: src/Syntax/Implicits.hs (hunk 2 above)
Comment (zardyh) [reply src/Syntax/Implicits.hs]: yes.

File: src/Backend/Lua/Builtin.hs (hunk 1 above)
Comment (zardyh) [reply src/Backend/Lua/Builtin.hs]: You come up with a better way to do it, then :/

File: tests/parser/fail_span.ml (hunk 3 above)
Comment (zardyh) [reply tests/parser/fail_span.ml]: sure

**Pull request merged**